x/dualstaking/keeper/hooks_test.go

```diff
@@ -157,3 +159,180 @@ func TestReDelegateToProvider(t *testing.T) {
+		require.Nil(t, err)
+	}
+
+	// unbond 25*5 tokens from validator
+	_, err = ts.TxUnbondValidator(delegatorAcc, validator, sdk.NewInt(25*5))
+	require.Nil(t, err)
+
+	res, err := ts.QueryDualstakingDelegatorProviders(delegator, false)
+	require.Nil(t, err)
+	for _, d := range res.Delegations {
+		switch d.Provider {
+		case providers[0]:
+			require.True(t, d.Amount.Amount.IsZero())
+		case providers[1]:
+			require.True(t, d.Amount.Amount.IsZero())
+		case providers[2]:
+			require.Equal(t, int64(20), d.Amount.Amount.Int64())
+		case providers[3]:
+			require.Equal(t, int64(30), d.Amount.Amount)
+		case providers[4]:
+			require.Equal(t, int64(35), d.Amount.Amount) // highest delegation is decreased by uniform amount + remainder
+		default:
+			require.FailNow(t, "unexpected provider in delegations")
+		}
+	}
+
+	diff, err := ts.Keepers.Dualstaking.VerifyDelegatorBalance(ts.Ctx, delegatorAcc.Addr)
+	require.Nil(t, err)
+	require.True(t, diff.IsZero())
+}
+
+// TestValidatorSlash checks that after a validator gets slashed, the delegators' providers
+// get slashed as expected (by an equal amount to preserve the validators-providers delegation balance)
+func TestValidatorSlash(t *testing.T) {
+	ts := newTester(t)
+	ts.addValidators(2)
+	ts.addProviders(5)
+	ts.addClients(1)
+	amount := sdk.NewIntFromUint64(10000)
+
+	// create validators and providers
+	var validators []sigs.Account
+	for i := 0; i < 2; i++ {
+		validator, _ := ts.GetAccount(common.VALIDATOR, i)
+		_, err := ts.TxCreateValidator(validator, amount)
+		require.Nil(t, err)
+		validators = append(validators, validator)
+	}
+
+	for i := 0; i < 5; i++ {
+		provider, _ := ts.GetAccount(common.PROVIDER, i)
+		err := ts.StakeProvider(provider.Addr.String(), ts.spec, amount.Int64())
+		require.Nil(t, err)
+	}
+
+	ts.AdvanceEpoch()
+
+	// delegate to validators (automatically delegates to empty provider)
+	delegatorAcc, delegator := ts.GetAccount(common.CONSUMER, 0)
+	_, err := ts.TxDelegateValidator(delegatorAcc, validators[0], sdk.NewInt(200))
+	require.Nil(t, err)
+	_, err = ts.TxDelegateValidator(delegatorAcc, validators[1], sdk.NewInt(10))
+	require.Nil(t, err)
+
+	// redelegate from empty provider to all providers with fixed amounts
+	redelegateAmts := []math.Int{
+		sdk.NewInt(10),
+		sdk.NewInt(20),
+		sdk.NewInt(50),
+		sdk.NewInt(60),
+		sdk.NewInt(70),
+	}
+	var providers []string
+	for i := 0; i < 5; i++ {
+		_, provider := ts.GetAccount(common.PROVIDER, i)
+		providers = append(providers, provider)
+		_, err = ts.TxDualstakingRedelegate(delegatorAcc.Addr.String(),
+			dualstakingkeeper.EMPTY_PROVIDER,
+			provider,
+			dualstakingkeeper.EMPTY_PROVIDER_CHAINID,
+			ts.spec.Index,
+			sdk.NewCoin(epochstoragetypes.TokenDenom, redelegateAmts[i]))
+		require.Nil(t, err)
+	}
+
+	// sanity check: redelegate from val2 to val1 and check delegations balance
+	_, err = ts.TxReDelegateValidator(delegatorAcc, validators[1], validators[0], sdk.NewInt(10))
```

Comment:
what i meant for the testing is to test redelegate from one provider to another, not from validator
so use txdualstakingredelegate instead

Reply:
done. see https://github.com/lavanet/lava/commit/117a1c20140e60fe7004324b1e4e2dfe7d720351